### PR TITLE
Use Java's LongConsumer

### PR DESCRIPTION
**Pull request by Maithem**
Since longlong.LongConsumer provides the same functionality as
java.util.function.LongConsumer I have replaced it. This allows
applications to reuse consumers without need to convert or wrap
another consumer that is functionally equivalent.

### CODE REVIEW

**Comment by Maithem** [comment]
@blacelle Question, is there a reason why `java.util.function.LongConsumer` in the first place ?

**Comment by blacelle** [comment]
@Maithem At first sight, I'm ashamed. At second sight, I decided to be consistant with `org.roaringbitmap.ImmutableBitmapDataProvider.forEach(IntConsumer)` and `org.roaringbitmap.IntConsumer`.

Hence, I would not merge this except if 32bits RoaringBitmap is impacted accordingly.

> This allows
applications to reuse consumers without need to convert or wrap
another consumer that is functionally equivalent.

With lambda, you do not need to rewrap:

`bitmap.forEach(standardLongConsumer::accept);`

**Comment by richardstartin** [comment]
This was added back in 2017 before we baselined on JDK8 so made sense. Note that this is a breaking change now, for any code where anonymous classes have been used.

**Comment by lemire** [comment]
If we break people's code, we should signal it by at least bumping the minor version at release time.

**Comment by richardstartin** [comment]
> If we break people's code, we should signal it by at least bumping the minor version at release time.

Exactly, which is why it's better to target `IntConsumer` at the same time otherwise you burn two minor versions. Doesn't need to be this PR, but I wouldn't expect to see this released without doing the other task.

**Comment by lemire** [comment]
@Maithem Would you consider also removing IntConsumer as well? Then I would hold this PR until we are ready to increment our minor version.

**Comment by Maithem** [comment]
Addressed comments.

File: examples/src/main/java/ForEachExample.java
Comment (blacelle) [inline examples/src/main/java/ForEachExample.java:28]: I don't get the need for `(IntConsumer)`

File: jmh/src/jmh/java/org/roaringbitmap/map/MapBenchmark.java
Comment (blacelle) [inline jmh/src/jmh/java/org/roaringbitmap/map/MapBenchmark.java:30]: I don't get the need for `(IntConsumer)`

File: jmh/src/jmh/java/org/roaringbitmap/map/MapBenchmark.java
Comment (blacelle) [inline jmh/src/jmh/java/org/roaringbitmap/map/MapBenchmark.java:39]: I don't get the need for `(IntConsumer)`

**Review by blacelle — CHANGES_REQUESTED** [review]
Irrelevant cast from lambda to `(IntConsumer)`

File: examples/src/main/java/ForEachExample.java
Comment (Maithem) [reply examples/src/main/java/ForEachExample.java:28]: It's an ambiguous call because `RoaringBitmap` implements both `Iterable::forEach(Consumer<? super T> action)` and  `ImmutableBitmapDataProvider::forEach(IntConsumer ic)` so it needs a cast.

File: jmh/src/jmh/java/org/roaringbitmap/map/MapBenchmark.java
Comment (Maithem) [reply jmh/src/jmh/java/org/roaringbitmap/map/MapBenchmark.java:30]: See https://github.com/RoaringBitmap/RoaringBitmap/pull/455/files#r571490784

**Review by richardstartin — APPROVED** [review]

File: examples/src/main/java/ForEachExample.java
Comment (blacelle) [reply examples/src/main/java/ForEachExample.java:28]: @lemire should we take this oportunity to change one of the method name? It looks a pity that a standard lambda is not applyable.

Comment (blacelle) [reply examples/src/main/java/ForEachExample.java:28]: or `implements Iterable<Integer>` should be dropped alltogether? Accepting a Consumer of autoboxed Integers looks legacy to me...

There is already org.roaringbitmap.IntIterator. We may leverage this signature change to propose a `.stream()` returning an `IntStream`.

**Review by blacelle — APPROVED** [review]
I approve it is a step forward, but I'm convinced we should take the oportunity to go one or 2 steps further (e.g. removing implements Iterable<Integer>).

**Comment by Maithem** [comment]
> Iterable

Dropping the `Iterable` will prevent users to use use the `RoaringBitmap` in enhanced for-Loops (i.e., `for (int i : bitmap);`).

File: examples/src/main/java/ForEachExample.java
Comment (Maithem) [reply examples/src/main/java/ForEachExample.java:28]: Dropping the `Iterable` will prevent users to use use the `RoaringBitmap` in enhanced for-Loops `(i.e., for (int i : bitmap);)`.



I actually tried `Consumer<Integer>` to remove the superfluous cast, but the trade off is syntax vs robustness. For boxed integers, nulls will be allowed (wont break compilation) and that can be problematic. You can see here: https://github.com/RoaringBitmap/RoaringBitmap/compare/master...Maithem:useJavasLongConsumer2

Comment (blacelle) [reply examples/src/main/java/ForEachExample.java:28]: > Dropping the Iterable will prevent users to use use the RoaringBitmap in enhanced for-Loops (i.e., for (int i : bitmap);).

Totally fine for me. Such could would go through an autoboxed Integers even if the `for` may suggests only primitive ints are considered.

> Consumer<Integer>

Again, I would feel this as legacy ; nulls being allowed being another sign of something wrong.

I feel we should not suggest any wrong usage of the library. Massive autoboxing looks very bad to me -> IntConsumer + IntStream are interfaces which did not exist when Iterable<Integer> was introduced.

**Comment by blacelle** [comment]
This is the point of my suggestion. `for (int i : bitmap);` would generate tons of autoboxed `Integers`.

**Comment by richardstartin** [comment]
I see Benoit’s point, implementing Iterable encourages users to do something not in their interests as well as creating ambiguity. I think implements Iterable should be removed, but perhaps in another PR. Maintainers of code broken by that change would probably be grateful in the end.

**Comment by Maithem** [comment]
Can we merge this PR ?

**Comment by lemire** [comment]
@Maithem I suggest we hold on to it until we are ready to release 1.0. 

It would be nice to sort out https://github.com/RoaringBitmap/RoaringBitmap/pull/456 before we do.

**Comment by Maithem** [comment]
@blacelle @richardstartin @lemire This PR drops the Iterable https://github.com/RoaringBitmap/RoaringBitmap/pull/461

**Comment by blacelle** [comment]
@lemire Could you prepare a 1.0-SNAPSHOT branch, to enable merging smaller PR without impacting master/coming releases? Thanks

**Comment by lemire** [comment]
There is now a 1.0-SNAPSHOT branch as requested.

**Comment by lemire** [comment]
I have changed the base for this PR to 1.0-SNAPSHOT.

**Pull request merged**
